src/librustc_typeck/check/closure.rs

```diff
@@ -269,4 +273,408 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+/// structurally, producing a new type R that will be used for the
+/// type in the final signature. One subtle point: both types may
+/// contain late-bound regions with depth 1: these late-bound regions
+/// will be bound by the closure signature.
+///
+/// If all the type pairs in the signatures are relatable, we are
+/// attempting to ensure two key invariants regarding the results:
+///
+/// - First, each type R only contains **bound regions that
+///   originated from the corresponding expected type E**.
+///   - That is, if the user wrote `|x: &str, y|`, the resulting bound
+///     region name in `x` (`Anon(0)`) would not be combined with some
+///     bound region name that appears in the expected type for `y`. The bound region
+///     names from the expected/supplied types effectively inherit distinct
+///     namespaces, and should never be mixed.
+/// - Second, any bound region appears that appears in some expected type E,
+///   also appears in the corresponding spot in the resulting type R.
+///   - In particular, imagine that we have an expected signature
+///     `(&str) -> &str`, but the user wrote `|x: &'a str| ...`. In
+///     the supplied type, the region `'a` is free.
+///     - We could choose to use the final signature `(&'a str) ->
+///       &str` without violating the first invariant, since we did
+///       "intermix" any bound region names from the supplied types
+///       with those of the expected type.
+///     - However, if we did so, we have lost the constraint expressed
+///       in the expected signature, that the return type must relate
+///       to the input.  In fact, this resulting signature is
+///       inconsistent, since a bound region cannot only appear in the
+///       input (this was #38714).
+///
+/// To produce the type R, we relate the types recursively, which
+/// requires that the types generally match. If that is not true, an
+/// error results, which will result in the expected signature being
+/// ignored. The places where the types do not have to match, and the
+/// resulting type in those scenarios, are as follows:
+///
+/// - If the expected type E is an (unbound) inference variable,
+///   and S is not an inference variable, then the resulting type will be S,
+///   so long as S does not contain anonymous regions.
+/// - If the supplied type S is an unbound inference variable (i.e.,
+///   the user wrote `_`), we will use E.
+/// - When we encounter types with internal binders, e.g. `fn(&str)`
+///   or `dyn Fn(&str)`, then we want to take the supplied type
+///   (justification below). To preserve the invariants listed above,
+///   we will error if either of the following is true (but otherwise
+///   use the supplied type):
+///   - the supplied type includes any regions bound at the closure depth
+///     - in this case, using the supplied type would inject names from the supplied
+///       side into the closure binder
```

Comment:
That takes some effort to create and I would like a test for it:

```Rust
type Foo<'a> = fn(&'a u32);

fn main() {
    foo(|_x: Foo| {});
}
```

Reply:
Similarly, replacing anonymous regions that occur multiple times with different regions would also allow this to compile (I would like a test):
```Rust
type Foo<'a, 'b> = &'a mut (&'a (), &'b ());
type Foo1<'a> = Foo<'a, 'a>;
fn give_foo<F: for<'a, 'b> Fn(Foo<'a, 'b>)>(_f: F) {}
fn assert_foo1(f: Foo1) {}
fn this_must_error() {
    give_foo(|f: Foo1| {
        assert_foo1(f) //~ ERROR
    });
}
fn but_this_doesnt() {
    give_foo(|f: Foo1| {
        // f
    });
}
```

Reply:
@arielb1 

> That takes some effort to create and I would like a test for it:

I had the following test:

```rust
fn expect_free_supply_free<'x>(x: &'x u32) {
    // Here, the type given for `'x` "obscures" a region from the
    // expected signature that is bound at closure level, so we don't
    // take anything from the expected type. You can see this in bits
    // of error message I have extracted.
    with_closure_expecting_fn_with_free_region(|x: fn(&'x u32), y| {});
    //~^ ERROR implements the trait `std::ops::FnOnce<(fn(&'x u32), _)>`
    //~| ERROR type mismatch
}
```

but I will add also the test you requested. (It errors.)

Reply:
Regarding your second test, it covers an interesting case I had not considered. I think that it is troubling that `but_this_doesnt` does not error. It seems like if the user wrote `Foo1` than they have "asserted" that they want the two regions to be the same. I suspect we want to treat repeated anonymous regions carefully -- i.e., rather like named regions?